R/get_predicted_args.R

```diff
@@ -0,0 +1,217 @@
+  ## TODO remove in a later update
+  # backward compatibility
+  if (identical(predict, "relation")) {
+    message(format_message(
+      '`predict = "relation" is deprecated.',
+      'Please use `predict = "expectation" instead.'
+    ))
+    predict <- "expectation"
+  }
+
+  # Warn if get_predicted() is not called with an easystats- or
+  # model-supported predicted type
+  easystats_methods <- c("expectation", "link", "prediction", "classification")
+  type_methods <- suppressWarnings(eval(formals(predict_method)$type))
+  supported <- c(easystats_methods, type_methods)
+  if (isTRUE(verbose) && !is.null(predict) && !predict %in% supported) {
+    msg <- format_message(
+      sprintf('`predict` = "%s"` is not officially supported by `get_predicted()`.', predict),
+      '`predict` will be passed directly to the `predict()` method for the model and not validated.',
+      'Please check the validity and scale of the results.',
+      'Set `verbose = FALSE` to silence this warning, or use one of the supported values for the `predict` argument:',
+      paste(" ", paste(sprintf('"%s"', setdiff(easystats_methods, c("expected", "predicted"))), collapse = ", "))
+    )
+    warning(msg, call. = FALSE)
+  }
+
+
+  transform <- FALSE
+
+  # type_arg  = what we pass own to type
+  # scale_arg = the scale of the predictions (link, response, terms, ...)
+  #             this may differ from type_arg, because we back-transform
+  # transform = whether we back-transform predictions, SE and CI
+
+  # define prediction type (that's for the initial call to stats::predict)
+
+  # terms
+  if (predict == "terms") {
+    type_arg <- "terms"
+    scale_arg <- "terms"
+    transform <- FALSE
+
+    # linear models are always on response scale (there is no other)
+  } else if (info$is_linear) {
+    type_arg <- "response"
+    scale_arg <- "response"
+    transform <- FALSE
+
+    # type = "response" always on link-scale - for later back-transformation
+  } else if (predict %in% c("expectation", "response", "prediction", "classification")) {
+
+    if (inherits(x, c("hurdle", "zeroinfl", "zerotrunc"))) {
+      # pscl: hurdle/zeroinfl and countreg
+      type_arg <- "count"
+    } else {
+      # default behaviour for "response"
+      type_arg <- "link"
+    }
+    scale_arg <- "response"
+    transform <- TRUE
+
+    # link-scale
+  } else if (predict == "link") {
+    type_arg <- "link"
+    scale_arg <- "link"
+    transform <- FALSE
+
+    # user provided a valid "type" value, which is not one of our "predict" values
+  } else if (predict %in% type_methods) {
+
+    if (predict == "count") {
+      # pscl
+      type_arg <- "count"
+      scale_arg <- "link"
+      transform <- FALSE
+
+    } else if (predict == "zero") {
+      # pscl
+      type_arg <- "zero"
+      scale_arg <- "link"
+      transform <- FALSE
+
+    } else {
+      # unknown / default
+      type_arg <- scale_arg <- predict
+      transform <- FALSE
+    }
+
+    # unknown
+  } else {
```

Comment:
Why check if `predict %in% type_methods` above if we're behaving the same way here anyway? Can we afford to be stricter and just reject unknown `predict` values? Are we confident that `type_methods` covers all the relevant values?

Reply:
This is some magic from @bwiernik:

```r
  # check `predict` user-input
  predict_method <- lapply(
    class(x), function(i) {
      tryCatch(utils::getS3method("predict", i),
               error = function(e) NULL)
    }
  )
  predict_method <- tryCatch(predict_method[!sapply(predict_method, is.null)][[1]], error = function(e) NULL)

  type_methods <- suppressWarnings(eval(formals(predict_method)$type))
```

so yes, `type_methods` should cover all possible types. It is rather that there might be valid `type` values that we did not cover yet, and these should be caught somewhere. We need to extend this if-else list when we add model by model.